File: internal/cloudez/payload_test.go
package cloudez

import (
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"sort"
	"strings"
	"testing"
	"time"
)

// escrever monta uma arvore a partir de um mapa caminho->conteudo. Um caminho
// terminado em "/" vira diretorio vazio.
func escrever(t *testing.T, arquivos map[string]string) string {
	t.Helper()
	dir := t.TempDir()
	for rel, conteudo := range arquivos {
		full := filepath.Join(dir, filepath.FromSlash(rel))
		if strings.HasSuffix(rel, "/") {
			if err := os.MkdirAll(full, 0o755); err != nil {
				t.Fatalf("mkdir %s: %v", rel, err)
			}
			continue
		}
		if err := os.MkdirAll(filepath.Dir(full), 0o755); err != nil {
			t.Fatalf("mkdir %s: %v", rel, err)
		}
		if err := os.WriteFile(full, []byte(conteudo), 0o644); err != nil {
			t.Fatalf("write %s: %v", rel, err)
		}
	}
	return dir
}

func hashDe(t *testing.T, dir string) PayloadHash {
	t.Helper()
	h, err := HashPayload(dir)
	if err != nil {
		t.Fatalf("HashPayload(%s): %v", dir, err)
	}
	return h
}

// O contrato central: o hash cobre exatamente o conjunto que o transfer envia.
// Se divergirem, o release_id descreve algo que nao foi ao ar — pior do que nao
// ter identificador. A regra de exclusao esta escrita duas vezes (aqui e no
// tarArgs do cmd/sync) e este teste e o que impede as duas de divergirem.
func TestHashCobreOMesmoConjuntoQueOTar(t *testing.T) {
	if _, err := exec.LookPath("tar"); err != nil {
		t.Skip("tar indisponivel")
	}

	dir := escrever(t, map[string]string{
		"index.html":            "<h1>oi</h1>",
		".gitignore":            "node_modules\n",
		".github/workflows/ci":  "on: push\n",
		".git/HEAD":             "ref: refs/heads/main\n",
		".git/objects/ab/cd":    "binario",
		"vendor/lib/.git/HEAD":  "ref: refs/heads/main\n",
		"vendor/lib/index.js":   "module.exports = 1\n",
		"assets/img/logo.svg":   "<svg/>",
		"docs/.gitkeep":         "",
		"nested/deep/a/b/c.txt": "c",
	})

	// Mesmos argumentos do transfer, menos a compressao: aqui interessa a lista.
	out, err := exec.Command("tar", "-cf", "-", "--exclude", ".git", "-C", dir, ".").Output()
	if err != nil {
		t.Fatalf("tar -c: %v", err)
	}
	listar := exec.Command("tar", "-tf", "-")
	listar.Stdin = strings.NewReader(string(out))
	listagem, err := listar.Output()
	if err != nil {
		t.Fatalf("tar -t: %v", err)
	}

	doTar := map[string]bool{}
	for _, linha := range strings.Split(strings.TrimSpace(string(listagem)), "\n") {
		linha = strings.TrimPrefix(strings.TrimSpace(linha), "./")
		// Diretorios saem com barra final; o hash so considera arquivos.
		if linha == "" || linha == "." || strings.HasSuffix(linha, "/") {
			continue
		}
		doTar[linha] = true
	}

	entries, err := payloadEntries(dir)
	if err != nil {
		t.Fatalf("payloadEntries: %v", err)
	}
	doHash := map[string]bool{}
	for _, e := range entries {
		doHash[e.rel] = true
	}

	for rel := range doTar {
		if !doHash[rel] {
			t.Errorf("o tar envia %q, mas o hash nao o cobre", rel)
		}
	}
	for rel := range doHash {
		if !doTar[rel] {
			t.Errorf("o hash cobre %q, mas o tar nao o envia", rel)
		}
	}

	if len(doHash) == 0 {
		t.Fatal("nenhum arquivo considerado — o teste nao verificou nada")
	}
}

// `.gitignore` e `.github/` precisam sobreviver ate o servidor: a exclusao e por
// nome puro, nao por prefixo. Ja o `.git` de um submodulo, em qualquer
// profundidade, cai — porque o padrao do tar nao e ancorado na raiz.
func TestExclusaoPegaGitEmQualquerNivelMasNaoOsVizinhos(t *testing.T) {
	casos := []struct {
		rel      string
		excluido bool
	}{
		{".git", true},
		{".git/HEAD", true},
		{"vendor/lib/.git", true},
		{"vendor/lib/.git/objects/ab", true},
		{".gitignore", false},
		{".github", false},
		{".github/workflows/ci.yml", false},
		{"src/.gitkeep", false},
		{"agit/x", false},
		{"x/.gitmodules", false},
	}
	for _, c := range casos {
		if got := excludedFromPayload(c.rel); got != c.excluido {
			t.Errorf("excludedFromPayload(%q) = %v, queria %v", c.rel, got, c.excluido)
		}
	}
}

// O identificador so serve se a mesma entrada der o mesmo hash em maquinas
// diferentes. A ordem da caminhada depende do sistema de arquivos, entao o
// mesmo conteudo criado em ordem diferente precisa colidir de proposito.
func TestMesmoConteudoMesmoHash(t *testing.T) {
	a := escrever(t, map[string]string{"a.txt": "um", "b/c.txt": "dois"})
	b := escrever(t, map[string]string{"b/c.txt": "dois", "a.txt": "um"})

	if hashDe(t, a).Sum != hashDe(t, b).Sum {
		t.Error("mesmo conteudo em ordem de criacao diferente deu hashes diferentes")
	}
}

// mtime muda a cada clone do repositorio sem que uma linha do projeto mude. Se
// entrasse no hash, o CI e a maquina do desenvolvedor nunca concordariam.
func TestMtimeNaoEntraNoHash(t *testing.T) {
	dir := escrever(t, map[string]string{"a.txt": "um"})
	antes := hashDe(t, dir)

	passado := time.Now().Add(-72 * time.Hour)
	if err := os.Chtimes(filepath.Join(dir, "a.txt"), passado, passado); err != nil {
		t.Fatalf("chtimes: %v", err)
	}

	if hashDe(t, dir).Sum != antes.Sum {
		t.Error("mudar o mtime mudou o hash")
	}
}

// O bit de execucao muda o comportamento do que esta no ar — um entrypoint que
// perde o +x quebra o container — entao precisa contar.
func TestBitDeExecucaoEntraNoHash(t *testing.T) {
	if runtime.GOOS == "windows" {
		t.Skip("modo de arquivo nao e comparavel no Windows")
	}

	dir := escrever(t, map[string]string{"entrypoint.sh": "#!/bin/sh\n"})
	semExec := hashDe(t, dir)

	if err := os.Chmod(filepath.Join(dir, "entrypoint.sh"), 0o755); err != nil {
		t.Fatalf("chmod: %v", err)
	}

	if hashDe(t, dir).Sum == semExec.Sum {
		t.Error("dar +x nao mudou o hash")
	}
}

// Sem um separador que nao pode aparecer num caminho, dois conjuntos distintos
// de arquivos concatenariam para os mesmos bytes. O caso classico: o nome de um
// invade o conteudo do outro.
func TestFronteiraEntreNomeEConteudo(t *testing.T) {
	a := escrever(t, map[string]string{"ab": "c"})
	b := escrever(t, map[string]string{"a": "bc"})

	if hashDe(t, a).Sum == hashDe(t, b).Sum {
		t.Error(`{"ab": "c"} e {"a": "bc"} colidiram`)
	}
}

// Mover conteudo entre arquivos, mantendo os bytes totais, tem de mudar o hash.
func TestConteudoMovidoEntreArquivosMudaOHash(t *testing.T) {
	a := escrever(t, map[string]string{"x.txt": "um", "y.txt": "dois"})
	b := escrever(t, map[string]string{"x.txt": "dois", "y.txt": "um"})

	if hashDe(t, a).Sum == hashDe(t, b).Sum {
		t.Error("trocar o conteudo entre dois arquivos nao mudou o hash")
	}
}

// O tar preserva symlinks em vez de segui-los. Seguir aqui hashearia bytes que
// nunca viajam, e um link quebrado — que o tar transporta sem reclamar — viraria
// erro de deploy.
func TestSymlinkEHasheadoPeloAlvo(t *testing.T) {
	if runtime.GOOS == "windows" {
		t.Skip("symlink exige privilegio no Windows")
	}

	dir := escrever(t, map[string]string{"real.txt": "conteudo"})
	if err := os.Symlink("real.txt", filepath.Join(dir, "link.txt")); err != nil {
		t.Fatalf("symlink: %v", err)
	}
	comLink := hashDe(t, dir)

	// Mesma arvore, mas o link vira arquivo comum com o conteudo do alvo.
	copia := escrever(t, map[string]string{"real.txt": "conteudo", "link.txt": "conteudo"})

	if comLink.Sum == hashDe(t, copia).Sum {
		t.Error("o symlink foi seguido: hasheou o conteudo do alvo em vez do caminho")
	}

	// E um link quebrado nao pode falhar: o tar o transporta.
	quebrado := escrever(t, map[string]string{"a.txt": "x"})
	if err := os.Symlink("nao-existe.txt", filepath.Join(quebrado, "orfao")); err != nil {
		t.Fatalf("symlink: %v", err)
	}
	if _, err := HashPayload(quebrado); err != nil {
		t.Errorf("link quebrado virou erro: %v", err)
	}
}

// Diretorio cujo unico conteudo e `.git` nao tem nada publicavel. Quem chama
// precisa conseguir distinguir isso de um deploy legitimo — e o que o Files == 0
// permite, e o cmd/sync usa para recusar.
func TestSoGitNaoTemArquivoPublicavel(t *testing.T) {
	dir := escrever(t, map[string]string{".git/HEAD": "ref: refs/heads/main\n"})

	if h := hashDe(t, dir); h.Files != 0 {
		t.Errorf("Files = %d, queria 0", h.Files)
	}
}

// Files e Bytes vao para o usuario junto do sha; se mentirem, mentem numa tela
// onde alguem decide se o deploy parece certo.
func TestContagemEBytesConferem(t *testing.T) {
	dir := escrever(t, map[string]string{
		"a.txt":   "12345",
		"b/c.txt": "678",
		".git/x":  "ignorado",
	})

	h := hashDe(t, dir)
	if h.Files != 2 {
		t.Errorf("Files = %d, queria 2", h.Files)
	}
	if h.Bytes != 8 {
		t.Errorf("Bytes = %d, queria 8", h.Bytes)
	}
}

// Âncora entre plataformas. As outras propriedades daqui são relativas ("mudar
// X muda o hash"), e todas continuariam valendo se o Windows produzisse uma
// família de hashes inteiramente diferente da do Linux.
//
// Isso não seria detalhe: o release_id sairia diferente para o mesmo conteúdo
// dependendo de quem publicou, e duas releases idênticas passariam por
// distintas — exatamente o que o identificador existe para evitar.
//
// O mesmo valor está no job `windows-build` do CI, que roda o binário no
// Windows sobre esta mesma árvore. Os dois mudam JUNTOS: se você alterar o
// formato do hash, o teste daqui quebra, e o do CI também.
func TestHashDeArvoreConhecida(t *testing.T) {
	const esperado = "03ff8bfa4d33e6cad50a7101e974ef34d6e8ccd51a77fbaacf97b7355cef01fc"

	dir := escrever(t, map[string]string{
		"index.html":      "<h1>oi</h1>",
		"assets/logo.svg": "svg",
		".git/HEAD":       "ref: x",
	})

	h := hashDe(t, dir)
	if h.Files != 2 {
		t.Fatalf("Files = %d, queria 2 (o .git fica de fora)", h.Files)
	}
	if h.Sum != esperado {
		t.Errorf("hash mudou de formato:\n  veio  %s\n  queria %s\n\n"+
			"Se a mudança foi intencional, atualize TAMBÉM o CONTENT_SHA_ESPERADO "+
			"do job windows-build em .github/workflows/test.yml.", h.Sum, esperado)
	}
}

// Short e o que vira sufixo do release_id.
func TestShortTemSeteDigitos(t *testing.T) {
	dir := escrever(t, map[string]string{"a.txt": "um"})
	h := hashDe(t, dir)

	if len(h.Short()) != 7 {
		t.Errorf("Short() = %q, queria 7 digitos", h.Short())
	}
	if !strings.HasPrefix(h.Sum, h.Short()) {
		t.Errorf("Short() = %q nao e prefixo de %q", h.Short(), h.Sum)
	}
}

// Diretorios vazios nao contam: o tar os transporta, mas depois de extraidos num
// diretorio de release eles nao se distinguem de ausencia.
func TestDiretorioVazioNaoMudaOHash(t *testing.T) {
	comum := map[string]string{"a.txt": "um"}
	sem := escrever(t, comum)

	com := escrever(t, map[string]string{"a.txt": "um", "vazio/": ""})

	if hashDe(t, sem).Sum != hashDe(t, com).Sum {
		t.Error("um diretorio vazio a mais mudou o hash")
	}

	// E a lista precisa continuar estavel.
	entries, err := payloadEntries(com)
	if err != nil {
		t.Fatalf("payloadEntries: %v", err)
	}
	nomes := make([]string, len(entries))
	for i, e := range entries {
		nomes[i] = e.rel
	}
	if !sort.StringsAreSorted(nomes) {
		t.Errorf("lista fora de ordem: %v", nomes)
	}
}

File: test/payload.test.mjs
// O identificador de conteúdo do payload (bin/_payload.mjs).
//
// Port da suíte que vivia em internal/cloudez/payload_test.go. Roda em
// `node --test`, embutido no Node 20 — o piso que o plugin já declara —, então
// não acrescenta dependência nenhuma à árvore.
//
// Por que não em bats, como o resto: aqui o que se afirma é sobre uma FUNÇÃO, não
// sobre um comando. As propriedades relativas ("mudar X muda o hash") precisam
// hashear duas árvores e comparar, e fazer isso por linha de comando seria
// exercitar o adaptador para testar a biblioteca. O comportamento do
// `cloudez-sync` continua coberto em test/adapters.bats.

import assert from "node:assert/strict"
import { chmodSync, mkdirSync, mkdtempSync, symlinkSync, utimesSync, writeFileSync } from "node:fs"
import { tmpdir } from "node:os"
import { dirname, join } from "node:path"
import { spawnSync } from "node:child_process"
import { test } from "node:test"

import { excluidoDoPayload, hashPayload, listarPayload, payloadEntries } from "../bin/_payload.mjs"

const NO_WINDOWS = process.platform === "win32"

/**
 * Monta uma árvore a partir de um mapa caminho->conteúdo. Um caminho terminado em
 * "/" vira diretório vazio.
 */
function escrever(arquivos) {
  const dir = mkdtempSync(join(tmpdir(), "cloudez-payload-"))
  for (const [rel, conteudo] of Object.entries(arquivos)) {
    const full = join(dir, rel)
    if (rel.endsWith("/")) {
      mkdirSync(full, { recursive: true })
      continue
    }
    mkdirSync(dirname(full), { recursive: true })
    writeFileSync(full, conteudo)
  }
  return dir
}

const sha = (dir) => hashPayload(dir).content_sha256

/**
 * O tar REAL, não o mock de test/mocks/. Estes testes afirmam sobre a listagem de
 * um pacote, e o mock não empacota nada — um teste que o usasse passaria vazio.
 */
function tarReal(args, stdin) {
  return spawnSync("tar", args, {
    env: { ...process.env, PATH: "/usr/bin:/bin" },
    input: stdin,
    maxBuffer: 64 * 1024 * 1024,
  })
}

// O contrato central: o hash cobre exatamente o conjunto que o transfer envia. Se
// divergirem, o release_id descreve algo que não foi ao ar — pior do que não ter
// identificador.
//
// A lista deixou de estar escrita duas vezes: o transfer passa ao tar a MESMA que
// o hash percorre, por `--null -T -`. O que este teste ainda guarda é que o tar
// empacota exatamente o que recebe — nem a mais (um `.` implícito reintroduzindo
// o que foi podado) nem a menos (nome com espaço ou acento perdido no caminho).
test("o tar empacota exatamente a lista que o hash cobre", { skip: NO_WINDOWS && "tar do Windows não confere aqui" }, () => {
  const dir = escrever({
    "index.html": "<h1>oi</h1>",
    ".gitignore": "node_modules\n",
    ".github/workflows/ci": "on: push\n",
    ".git/HEAD": "ref: refs/heads/main\n",
    ".git/objects/ab/cd": "binario",
    "vendor/lib/.git/HEAD": "ref: refs/heads/main\n",
    "vendor/lib/index.js": "module.exports = 1\n",
    "assets/img/logo.svg": "<svg/>",
    "docs/.gitkeep": "",
    "nested/deep/a/b/c.txt": "c",
    "com espaço e acento.txt": "ç",
  })

  const { entries } = listarPayload(dir)
  const lista = entries.map((e) => e.rel).join("\n") + "\n"

  // Mesma invocação do transfer, menos a compressão: aqui interessa a lista.
  const pacote = tarReal(["-cf", "-", "-C", dir, "-T", "-"], lista)
  assert.equal(pacote.status, 0, `tar -c falhou: ${pacote.stderr}`)

  // A comparação é por EXTRAÇÃO, e não pela listagem do `tar -t`: o bsdtar
  // escapa não-ASCII em octal ali, e o teste passaria a afirmar sobre o formato
  // de exibição do tar em vez do conteúdo do pacote.
  const destino = mkdtempSync(join(tmpdir(), "cloudez-extraido-"))
  const extracao = tarReal(["-xf", "-", "-C", destino], pacote.stdout)
  assert.equal(extracao.status, 0, `tar -x falhou: ${extracao.stderr}`)

  // Normalizado dos dois lados: o bsdtar do macOS converte a forma Unicode do
  // nome ao empacotar (NFC/NFD), e sem isto o teste afirmaria sobre a
  // normalização da plataforma em vez de sobre o conjunto de arquivos.
  const norm = (r) => r.normalize("NFC")
  const doTar = new Set(payloadEntries(destino).map((e) => norm(e.rel)))
  const doHash = new Set(entries.map((e) => norm(e.rel)))

  for (const rel of doTar) assert.ok(doHash.has(rel), `o tar envia ${rel}, mas o hash não o cobre`)
  for (const rel of doHash) assert.ok(doTar.has(rel), `o hash cobre ${rel}, mas o tar não o envia`)
  assert.ok(doHash.size > 0, "nenhum arquivo considerado — o teste não verificou nada")
})

// `.gitignore` e `.github/` precisam sobreviver até o servidor: a exclusão é por
// nome puro, não por prefixo. Já o `.git` de um submódulo, em qualquer
// profundidade, cai — porque o padrão do tar não é ancorado na raiz.
test("a exclusão pega .git em qualquer nível, mas não os vizinhos", () => {
  const casos = [
    [".git", true],
    [".git/HEAD", true],
    ["vendor/lib/.git", true],
    ["vendor/lib/.git/objects/ab", true],
    [".gitignore", false],
    [".github", false],
    [".github/workflows/ci.yml", false],
    ["src/.gitkeep", false],
    ["agit/x", false],
    ["x/.gitmodules", false],
  ]
  for (const [rel, excluido] of casos) {
    assert.equal(excluidoDoPayload(rel), excluido, `excluidoDoPayload(${rel})`)
  }
})

// O identificador só serve se a mesma entrada der o mesmo hash em máquinas
// diferentes. A ordem da caminhada depende do sistema de arquivos, então o mesmo
// conteúdo criado em ordem diferente precisa colidir de propósito.
test("mesmo conteúdo, mesmo hash", () => {
  const a = escrever({ "a.txt": "um", "b/c.txt": "dois" })
  const b = escrever({ "b/c.txt": "dois", "a.txt": "um" })
  assert.equal(sha(a), sha(b))
})

// mtime muda a cada clone do repositório sem que uma linha do projeto mude. Se
// entrasse no hash, o CI e a máquina do desenvolvedor nunca concordariam.
test("mtime não entra no hash", () => {
  const dir = escrever({ "a.txt": "um" })
  const antes = sha(dir)

  const passado = new Date(Date.now() - 72 * 60 * 60 * 1000)
  utimesSync(join(dir, "a.txt"), passado, passado)

  assert.equal(sha(dir), antes, "mudar o mtime mudou o hash")
})

// O bit de execução muda o comportamento do que está no ar — um entrypoint que
// perde o +x quebra o container — então precisa contar.
test("o bit de execução entra no hash", { skip: NO_WINDOWS && "modo de arquivo não é comparável no Windows" }, () => {
  const dir = escrever({ "entrypoint.sh": "#!/bin/sh\n" })
  const semExec = sha(dir)

  chmodSync(join(dir, "entrypoint.sh"), 0o755)

  assert.notEqual(sha(dir), semExec, "dar +x não mudou o hash")
})

// Sem um separador que não pode aparecer num caminho, dois conjuntos distintos de
// arquivos concatenariam para os mesmos bytes. O caso clássico: o nome de um
// invade o conteúdo do outro.
test("a fronteira entre nome e conteúdo é respeitada", () => {
  const a = escrever({ ab: "c" })
  const b = escrever({ a: "bc" })
  assert.notEqual(sha(a), sha(b), '{"ab": "c"} e {"a": "bc"} colidiram')
})

test("conteúdo movido entre arquivos muda o hash", () => {
  const a = escrever({ "x.txt": "um", "y.txt": "dois" })
  const b = escrever({ "x.txt": "dois", "y.txt": "um" })
  assert.notEqual(sha(a), sha(b))
})

// O tar preserva symlinks em vez de segui-los. Seguir aqui hashearia bytes que
// nunca viajam, e um link quebrado — que o tar transporta sem reclamar — viraria
// erro de deploy.
test("o symlink é hasheado pelo alvo", { skip: NO_WINDOWS && "symlink exige privilégio no Windows" }, () => {
  const dir = escrever({ "real.txt": "conteudo" })
  symlinkSync("real.txt", join(dir, "link.txt"))
  const comLink = sha(dir)

  // Mesma árvore, mas o link vira arquivo comum com o conteúdo do alvo.
  const copia = escrever({ "real.txt": "conteudo", "link.txt": "conteudo" })
  assert.notEqual(comLink, sha(copia), "o symlink foi seguido: hasheou o conteúdo do alvo")

  // E um link quebrado não pode falhar: o tar o transporta.
  const quebrado = escrever({ "a.txt": "x" })
  symlinkSync("nao-existe.txt", join(quebrado, "orfao"))
  assert.doesNotThrow(() => hashPayload(quebrado), "link quebrado virou erro")
})

// Diretório cujo único conteúdo é `.git` não tem nada publicável. Quem chama
// precisa distinguir isso de um deploy legítimo — é o que o files === 0 permite, e
// o cloudez-sync usa para recusar.
test("só .git não tem arquivo publicável", () => {
  const dir = escrever({ ".git/HEAD": "ref: refs/heads/main\n" })
  assert.equal(hashPayload(dir).files, 0)
})

// files e bytes vão para o usuário junto do sha; se mentirem, mentem numa tela
// onde alguém decide se o deploy parece certo.
test("a contagem e os bytes conferem", () => {
  const dir = escrever({ "a.txt": "12345", "b/c.txt": "678", ".git/x": "ignorado" })
  const h = hashPayload(dir)
  assert.equal(h.files, 2)
  assert.equal(h.bytes, 8)
})

// Âncora entre plataformas. As outras propriedades daqui são relativas ("mudar X
// muda o hash"), e todas continuariam valendo se o Windows produzisse uma família
// de hashes inteiramente diferente da do Linux.
//
// Isso não seria detalhe: o release_id sairia diferente para o mesmo conteúdo
// dependendo de quem publicou, e duas releases idênticas passariam por distintas —
// exatamente o que o identificador existe para evitar.
//
// Este valor é o MESMO que a suíte em Go fixava antes do port. Que ele tenha
// sobrevivido à troca de linguagem é o que garante que nenhuma release já
// publicada mudou de identificador.
test("o hash de uma árvore conhecida não mudou", () => {
  const esperado = "03ff8bfa4d33e6cad50a7101e974ef34d6e8ccd51a77fbaacf97b7355cef01fc"

  const dir = escrever({
    "index.html": "<h1>oi</h1>",
    "assets/logo.svg": "svg",
    ".git/HEAD": "ref: x",
  })

  const h = hashPayload(dir)
  assert.equal(h.files, 2, "o .git tem de ficar de fora")
  assert.equal(h.content_sha256, esperado, "o formato do hash mudou — releases já publicadas mudariam de id")
})

// O prefixo de 7 dígitos é o que vira sufixo do release_id (ver commands/deploy.md).
test("os 7 primeiros dígitos servem de sufixo do release_id", () => {
  const dir = escrever({ "a.txt": "um" })
  const h = hashPayload(dir)
  const curto = h.content_sha256.slice(0, 7)
  assert.equal(curto.length, 7)
  assert.ok(h.content_sha256.startsWith(curto))
})

// Diretórios vazios não contam: o tar os transporta, mas depois de extraídos num
// diretório de release eles não se distinguem de ausência.
test("diretório vazio não muda o hash", () => {
  const sem = escrever({ "a.txt": "um" })
  const com = escrever({ "a.txt": "um", "vazio/": "" })
  assert.equal(sha(sem), sha(com), "um diretório vazio a mais mudou o hash")

  // E a lista precisa continuar estável.
  const nomes = payloadEntries(com).map((e) => e.rel)
  const ordenados = [...nomes].sort((a, b) => Buffer.compare(Buffer.from(a), Buffer.from(b)))
  assert.deepEqual(nomes, ordenados, `lista fora de ordem: ${nomes}`)
})

// ── .gitignore e .cloudezignore ──────────────────────────────────────────────
//
// O tar excluía só o `.git`, e o diretório publicado é o contexto de build: num
// projeto Node ele levava `node_modules` e a saída do build — 532 MB e 38 mil
// arquivos num site cujo fonte tem 0,3 MB, refeitos pelo build no servidor de
// qualquer jeito.

const rels = (dir) => listarPayload(dir).entries.map((e) => e.rel).sort()

test("o .gitignore poda, e não desce no que podou", () => {
  const dir = escrever({
    ".gitignore": "node_modules\n.next\n",
    "index.js": "1",
    "node_modules/pacote/index.js": "2",
    "node_modules/pacote/sub/mais.js": "3",
    ".next/build.js": "4",
  })

  assert.deepEqual(rels(dir), [".gitignore", "index.js"])

  const { ignore } = listarPayload(dir)
  // Conta os caminhos PODADOS, não os arquivos dentro deles: é o que torna a
  // exclusão barata — `node_modules` custa uma comparação, não 38 mil.
  assert.deepEqual(ignore.paths, [".next", "node_modules"])
  assert.deepEqual(ignore.sources, [".gitignore"])
})

test("o .cloudezignore também poda", () => {
  const dir = escrever({ ".cloudezignore": "temp/\n", "a.js": "1", "temp/x": "2" })
  assert.deepEqual(rels(dir), [".cloudezignore", "a.js"])
})

// A ordem de leitura é o que dá ao .cloudezignore a palavra final: ele resolve o
// caso do artefato ignorado pelo git que PRECISA ir ao ar.
test("o .cloudezignore desfaz o .gitignore com !", () => {
  const dir = escrever({
    ".gitignore": "dist\n",
    ".cloudezignore": "!dist\n",
    "dist/app.js": "1",
    "src/a.js": "2",
  })

  assert.ok(rels(dir).includes("dist/app.js"), "o ! do .cloudezignore devia ter reabilitado")
})

// Decisão registrada: o .dockerignore descreve o que não entra no contexto
// enviado ao daemon, e ali é CORRETO excluir o Dockerfile e o compose — o Docker
// os recebe por outro caminho. O deploy publica o diretório para construir
// depois, no servidor: respeitá-lo faria todo site em container falhar com
// compose_missing.
test("REGRESSÃO: o .dockerignore NÃO é lido", () => {
  const dir = escrever({
    ".dockerignore": "Dockerfile\ndocker-compose.yml\nnode_modules\n",
    Dockerfile: "FROM node",
    "docker-compose.yml": "services: {}",
    "node_modules/x/y.js": "1",
  })

  const lista = rels(dir)
  assert.ok(lista.includes("Dockerfile"), "sem o Dockerfile o build no servidor não existe")
  assert.ok(lista.includes("docker-compose.yml"), "sem o compose o deploy falha com compose_missing")
  assert.ok(lista.includes("node_modules/x/y.js"), "e nem a regra inofensiva vale — o arquivo não é lido")
  assert.deepEqual(listarPayload(dir).ignore.sources, [])
})

test("a âncora com barra limita à raiz", () => {
  const dir = escrever({
    ".gitignore": "/build\n",
    "build/a": "1",
    "pacote/build/b": "2",
  })

  assert.deepEqual(rels(dir), [".gitignore", "pacote/build/b"])
})

test("sem âncora, vale em qualquer profundidade", () => {
  const dir = escrever({ ".gitignore": "build\n", "build/a": "1", "pacote/build/b": "2", "c.js": "3" })
  assert.deepEqual(rels(dir), [".gitignore", "c.js"])
})

test("a barra no fim casa só diretório", () => {
  const dir = escrever({ ".gitignore": "build/\n", build: "sou um arquivo", "outro/build/a": "1" })
  assert.deepEqual(rels(dir), [".gitignore", "build"])
})

test("os curingas", () => {
  const dir = escrever({
    ".gitignore": "*.log\ntmp?\nsrc/**/gerado.js\n",
    "erro.log": "1",
    "tmp1/a": "2",
    "tmpXY/b": "3",
    "src/a/b/gerado.js": "4",
    "src/mantido.js": "5",
  })

  assert.deepEqual(rels(dir), [".gitignore", "src/mantido.js", "tmpXY/b"])
})

test("comentário e linha em branco não viram regra", () => {
  const dir = escrever({ ".gitignore": "# nada\n\n   \n", "a.js": "1" })
  assert.deepEqual(rels(dir), [".gitignore", "a.js"])
})

// O `.git` sai por uma regra própria, avaliada ANTES dos padrões — um `!` não o
// traz de volta. Publicar o histórico do repositório numa imagem não é escolha
// que se ofereça.
test("nem um ! reabilita o .git", () => {
  const dir = escrever({ ".cloudezignore": "!.git\n", ".git/HEAD": "ref", "a.js": "1" })
  assert.deepEqual(rels(dir), [".cloudezignore", "a.js"])
})

// Igual ao git: negação não reabilita arquivo sob diretório excluído, porque o
// diretório nem chega a ser percorrido.
test("o ! não reabilita arquivo dentro de diretório podado", () => {
  const dir = escrever({ ".gitignore": "dist\n!dist/importante.js\n", "dist/importante.js": "1", "a.js": "2" })
  assert.deepEqual(rels(dir), [".gitignore", "a.js"])
})

test("sem arquivo de padrão, o retorno não ganha campo", () => {
  const dir = escrever({ "a.js": "1" })
  assert.equal(hashPayload(dir).ignored, undefined)
  assert.deepEqual(listarPayload(dir).ignore.sources, [])
})

// Um `.gitignore` que exclua tudo deixaria o deploy publicar uma release vazia
// com identificador de aparência normal — o cloudez-sync recusa isso.
test("padrão que exclui tudo deixa zero arquivos publicáveis", () => {
  const dir = escrever({ ".cloudezignore": "*\n", "a.js": "1", "b.js": "2" })
  assert.equal(listarPayload(dir).entries.length, 0)
})

// ── nomes que o transporte não sabe carregar ─────────────────────────────────
//
// A lista vai ao tar separada por QUEBRA DE LINHA, e não por NUL: o `--null`
// existe no bsdtar e no GNU tar, mas não no busybox — que é o tar de qualquer
// imagem Alpine, base comum de CI para projeto Node.
//
// O preço são três caracteres, e recusá-los é a escolha: um arquivo que não vai
// junto produziria uma release incompleta com identificador de aparência normal.

const recusa = (dir) => {
  try {
    listarPayload(dir)
    return null
  } catch (e) {
    return e
  }
}

test("barra invertida no nome é recusada", { skip: NO_WINDOWS && "nome com barra invertida não existe no Windows" }, () => {
  // Medido no GNU tar: `com\tbarra.txt` numa lista sem --null vira uma busca por
  // um nome com TAB, e devolve "Cannot stat".
  const e = recusa(escrever({ "com\\tbarra.txt": "x", "ok.txt": "y" }))
  assert.ok(e, "devia recusar")
  assert.equal(e.code, "filename_unsupported")
  assert.match(e.message, /com\\tbarra\.txt/)
})

test("quebra de linha no nome é recusada", { skip: NO_WINDOWS && "nome com \\n não existe no Windows" }, () => {
  // Partiria a lista em dois nomes, e nenhum dos dois existe.
  const e = recusa(escrever({ "duas\nlinhas.txt": "x" }))
  assert.ok(e, "devia recusar")
  assert.equal(e.code, "filename_unsupported")
})

// Só os três. Espaço e aspas atravessam a lista sem escape nenhum — conferido no
// GNU tar —, e recusá-los seria barrar nome de arquivo perfeitamente comum.
test("espaço e aspas no nome PASSAM", () => {
  const dir = escrever({ "com espaço.txt": "x", 'com "aspas".txt': "y" })
  assert.equal(recusa(dir), null)
  assert.equal(listarPayload(dir).entries.length, 2)
})

// A recusa vem antes de qualquer byte viajar: quem falha é a listagem, não o tar.
test("a recusa não depende de o tar existir", { skip: NO_WINDOWS && "idem" }, () => {
  const e = recusa(escrever({ "x\\y.txt": "1" }))
  assert.equal(e.code, "filename_unsupported")
  assert.match(e.message, /Renomeie/)
})
